### PR TITLE
GAME: update and delete options

### PR DESCRIPTION
**Pull request by Adelheid483**
Issue link - https://github.com/TexnikBy/BGS/issues/11

### CODE REVIEW

File: server/src/Core/Interfaces/IGameNameDuplicationsChecker.cs [hunk 1]
```diff
@@ -0,0 +1,9 @@
+﻿using System.Threading;
+using System.Threading.Tasks;
+
+namespace BGS.ApplicationCore.Interfaces;
+
+public interface IGameNameDuplicationsChecker
```
Comment (TexnikBy) [inline server/src/Core/Interfaces/IGameNameDuplicationsChecker.cs]: rename to `IGameDuplicationChecker`.
In this case I don't think we should focus on implementation details in the interface name

File: server/src/Core/Interfaces/IGameNameDuplicationsChecker.cs [hunk 2]
```diff
@@ -0,0 +1,9 @@
...
+{
+    Task<bool> Check(string gameName, CancellationToken cancellationToken);
```
Comment (TexnikBy) [inline server/src/Core/Interfaces/IGameNameDuplicationsChecker.cs]: Pass the cancellationToken is optional, in this case I think we can leave it out

File: server/src/Core/Specifications/GameKeyByIdSpecification.cs
Comment (TexnikBy) [inline server/src/Core/Specifications/GameKeyByIdSpecification.cs:1]: this is bug?

File: server/src/Core/Specifications/GameNameDuplicationsSpecification.cs [hunk 3]
```diff
@@ -0,0 +1,12 @@
+﻿using Ardalis.Specification;
+using BGS.ApplicationCore.Entities;
+
+namespace BGS.ApplicationCore.Specifications;
+
+public sealed class GameNameDuplicationsSpecification : SingleResultSpecification<Game, bool>
```
Comment (TexnikBy) [inline server/src/Core/Specifications/GameNameDuplicationsSpecification.cs]: This specification is placed in the core module. Let's rename it to a more "common" name `GameByNameSpecification`.

File: server/src/Core/Validators/GameNameDuplicationsChecker.cs [hunk 4]
```diff
@@ -0,0 +1,18 @@
...
+
+namespace BGS.ApplicationCore.Validators;
+
+public class GameNameDuplicationsChecker(IRepository<Game> gameRepository) : IGameNameDuplicationsChecker
+{
+    public async Task<bool> Check(string gameName, CancellationToken cancellationToken)
+    {
+        return await gameRepository.AnyAsync(
+            new GameNameDuplicationsSpecification(gameName),
+            cancellationToken);
```
Comment (TexnikBy) [inline server/src/Core/Validators/GameNameDuplicationsChecker.cs]: don't forget about it - https://cezarypiatek.github.io/post/async-analyzers-p1/#1-redundant-asyncawait

File: server/src/Core/Validators/GameNameDuplicationsChecker.cs [hunk 5]
```diff
@@ -0,0 +1,18 @@
+﻿using System.Threading;
+using System.Threading.Tasks;
+using BGS.ApplicationCore.Entities;
+using BGS.ApplicationCore.Interfaces;
+using BGS.ApplicationCore.Specifications;
+using BGS.SharedKernel;
+
+namespace BGS.ApplicationCore.Validators;
```
Comment (TexnikBy) [inline server/src/Core/Validators/GameNameDuplicationsChecker.cs]: this checker relates to games :)

File: server/src/Core/Specifications/GameNameDuplicationsSpecification.cs [hunk 6]
```diff
@@ -0,0 +1,12 @@
...
+{
+    public GameNameDuplicationsSpecification(string gameName)
+    {
+        Query.Where(game => string.Equals(game.Name.ToUpper(), gameName.ToUpper()));
```
Comment (TexnikBy) [inline server/src/Core/Specifications/GameNameDuplicationsSpecification.cs]: Let's think about our current situation. There cannot be two or more games with the same name in our application. So we need to configure restriction at the db level. Let's change type of name to `citext` and remove `.ToUpper()`

File: server/src/UseCases/Games/Update/UpdateGameModel.cs [hunk 7]
```diff
@@ -0,0 +1,10 @@
+using System;
+
+namespace BGS.UseCases.Games.Update;
+
+public class UpdateGameModel
```
Comment (TexnikBy) [inline server/src/UseCases/Games/Update/UpdateGameModel.cs]: record

File: server/tests/Tests.Core/Tests.Core.csproj [hunk 8]
```diff
@@ -0,0 +1,34 @@
+<Project Sdk="Microsoft.NET.Sdk">
+
+    <PropertyGroup>
+        <TargetFramework>net8.0</TargetFramework>
+        <ImplicitUsings>enable</ImplicitUsings>
+        <Nullable>disable</Nullable>
```
Comment (TexnikBy) [inline server/tests/Tests.Core/Tests.Core.csproj]: It's necessary?

File: server/tests/Tests.Core/Tests.Core.csproj [hunk 9]
```diff
@@ -0,0 +1,34 @@
...
+
+        <IsPackable>false</IsPackable>
+        <IsTestProject>true</IsTestProject>
+        <RootNamespace>BGS.Tests.Core</RootNamespace>
+        <OutputType>Library</OutputType>
```
Comment (TexnikBy) [inline server/tests/Tests.Core/Tests.Core.csproj]: It's necessary?

File: server/tests/Tests.Core/Tests.Core.csproj [hunk 10]
```diff
@@ -0,0 +1,34 @@
...
+        <PackageReference Include="Microsoft.NET.Test.Sdk" Version="17.6.0"/>
+        <PackageReference Include="xunit" Version="2.4.2"/>
+        <PackageReference Include="xunit.runner.visualstudio" Version="2.4.5">
+            <IncludeAssets>runtime; build; native; contentfiles; analyzers; buildtransitive</IncludeAssets>
+            <PrivateAssets>all</PrivateAssets>
+        </PackageReference>
+        <PackageReference Include="coverlet.collector" Version="6.0.0">
+            <IncludeAssets>runtime; build; native; contentfiles; analyzers; buildtransitive</IncludeAssets>
+            <PrivateAssets>all</PrivateAssets>
+        </PackageReference>
```
Comment (TexnikBy) [inline server/tests/Tests.Core/Tests.Core.csproj]: add `ProjectReference` to `Tests.Infrastructure.csproj` and remove unnecessary PackageReference except Microsoft.NET.Test.Sdk. Update Microsoft.NET.Test.Sdk to 17.8.0 version

**Review by TexnikBy — CHANGES_REQUESTED** [review]

File: server/src/Core/Interfaces/IGameNameDuplicationsChecker.cs (hunk 2 above)
Comment (Adelheid483) [reply server/src/Core/Interfaces/IGameNameDuplicationsChecker.cs]: removed

File: server/src/UseCases/Games/Create/CreateGameCommandHandler.cs [hunk 11]
```diff
@@ -1,29 +1,31 @@
...
-    public CreateGameCommandHandler(IRepository<Game> gameRepository)
-    {
-        _gameRepository = gameRepository;
-    }
-
     public async Task<Result> Handle(CreateGameCommand command, CancellationToken cancellationToken)
     {
+        if (await gameDuplicationChecker.Check(command.GameName))
+        {
+            return Result<Result>.Error($"Game with the name {command.GameName} is already exists.");
```
Comment (TexnikBy) [inline server/src/UseCases/Games/Create/CreateGameCommandHandler.cs]: this line is duplicated here and in `UpdateGameCommandHandler`. Let's change return type of `IGameDuplicationChecker` to Result, and don't use `Result<Result>` just create error method in Result.

**Pull request merged**